src/vm/types.rs

```diff
@@ -70,143 +212,399 @@ impl TupleTypeSignature {
-            Value::List(_v, type_signature) => type_signature.clone(),
-            Value::Tuple(v) => TypeSignature::new(AtomTypeIdentifier::TupleType(
-                v.type_signature.clone()), 0)
+        if let Value::List(_, type_signature) = x {
+            type_signature.clone()
+        } else {
+            let atom = match x {
+                Value::Void => AtomTypeIdentifier::VoidType,
+                Value::Principal(_,_) => AtomTypeIdentifier::PrincipalType,
+                Value::Int(_v) => AtomTypeIdentifier::IntType,
+                Value::Bool(_v) => AtomTypeIdentifier::BoolType,
+                Value::Buffer(buff_data) => AtomTypeIdentifier::BufferType(buff_data.length),
+                Value::Tuple(v) => AtomTypeIdentifier::TupleType(
+                    v.type_signature.clone()),
+                Value::List(_,_) => panic!("Unreachable code")
+            };
+
+            TypeSignature::new_atom(atom)
         }
     }
 
-    pub fn get_list_type_for(x: &Value) -> Result<TypeSignature, Error> {
-        match x {
-            Value::Void => Err(Error::InvalidArguments("Cannot construct list of void types".to_string())),
-            Value::Tuple(_a) => Err(Error::InvalidArguments("Cannot construct list of tuple types".to_string())),
-            _ => {
-                let mut base_type = TypeSignature::type_of(x);
-                base_type.dimension += 1;
-                Ok(base_type)
+    fn expand_to_admit(&mut self, x_type: &TypeSignature) -> Result<()> {
+        if let Some(ref mut my_list_dimensions) = self.list_dimensions {
+            if let Some(ref x_list_dimensions) = x_type.list_dimensions {
+                if my_list_dimensions.dimension != x_list_dimensions.dimension {
+                    return Err(Error::BadTypeConstruction)
+                }
+                if my_list_dimensions.max_len < x_list_dimensions.max_len {
+                    my_list_dimensions.max_len = x_list_dimensions.max_len;
+                }
+            } else {
+                return Err(Error::BadTypeConstruction)
+            }
+        } else {
+            if let Some(_) = x_type.list_dimensions {
+                return Err(Error::BadTypeConstruction)
+            }
+        }
+
+        self.atomic_type.expand_to_admit(& x_type.atomic_type)
+    }
+
+    // Checks if resulting type signature is of valid size.
+    // Aaron:
+    //    currently, this does "loose admission" for higher-order lists --
+    //     but should it do the same for buffers and tuples or is it better
+    //     like it is now, where it requires an exact type match on those?
+    //     e.g.: (list "abcd" "abc") will currently error because one etry is
+    //           if type (buffer 4) and the other is of type (buffer 3)
+    //       my feeling is that this should probably be allowed, and the resulting
+    //       type should be (list 2 (buffer 4)) 
```

Comment:
I agree, just as long as we can make it clear somehow that `(list "abcdefghijk" "abc")` costs as much to process as `(list "abcdefghijk" "        abc")`.

Reply:
yep -- we'd need to make this clear. I think the behavior we want is that for any functions which accepts two buffers, to treat the smaller buffer as left padded with zeros. currently we don't have _any_ buffer functions (except `eq?`).

Reply:
How do you want to handle `eq?` for buffers that get implicitly left-padded with zeros?  I don't think `\x00abc` and `abc` are necessarily equal.  Maybe the buffer type should include a `length` field internally to avoid this problem?

Reply:
Yeah, I was just going to comment on this. Those shouldn't be equal. Buffers should have a length, which they do currently (via their `vec.len()`), but the Buffer type has a _maximum length_, which is a little silly for a buffer's runtime type (because its length is always equal to its maximum length), but for declared types like public function inputs and datamap keys, values, it makes sense, because the maximum length determines whether or not a given runtime value would be admissable.

Reply:
(this contradicts my previous response up there --- I don't think we should be automatically extending buffers. if we decide to include buffer functions like `strncat`, they should specify their padding behavior, if they have any)